tests/test_compile_kwargs.py

```diff
@@ -0,0 +1,563 @@
+from tensorflow.keras import metrics as metrics_module
+from tensorflow.keras import optimizers as optimizers_module
+from tensorflow.keras.layers import Dense, Input
+from tensorflow.keras.models import Model
+from tensorflow.python.keras.engine.compile_utils import (
+    losses_mod,
+    metrics_mod,
+)
+
+from scikeras.wrappers import KerasClassifier
+
+
+def get_model(num_hidden=10, meta=None, compile_kwargs=None):
+    inp = Input(shape=(meta["n_features_in_"],))
+    hidden = Dense(num_hidden, activation="relu")(inp)
+    out = [
+        Dense(1, activation="sigmoid", name=f"out{i+1}")(hidden)
+        for i in range(meta["n_outputs_"])
+    ]
+    model = Model(inp, out)
+    model.compile(**compile_kwargs)
+    return model
+
+
+@pytest.mark.parametrize("optimizer", (optimizers_module.SGD, "SGD", "sgd"))
+@pytest.mark.parametrize("n_outputs_", (1, 2))
+def test_optimizer(optimizer, n_outputs_):
+    """Tests compiling of single optimizer with options.
+    Since there can only ever be a single optimizer, there is no
+    ("name", optimizer, "output") option.
+    Only optimizer classes will be compiled with custom options,
+    all others (class names, function names) should pass through
+    untouched.
+    """
+    # Single output
+    X, y = make_classification()
+    y = np.column_stack([y for _ in range(n_outputs_)]).squeeze()
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        optimizer__learning_rate=0.15,
+        optimizer__momentum=0.5,
+        loss="binary_crossentropy",
+    )
+    est.fit(X, y)
+    if isclass(optimizer):
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.momentum.value().numpy(), 0.5, decimal=2
+        )
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.learning_rate.value().numpy(), 0.15, decimal=2
+        )
+    else:
+        # TF actually instantiates an optimizer object even if passed a string
+        # Checking __class__ is not a universally valid check, but it works for the
+        # test optimizer (SGD)
+        assert (
+            est.model_.optimizer.__class__
+            == optimizers_module.get(optimizer).__class__
+        )
+
+
+def test_optimizer_invalid_string():
+    """Tests that a ValueError is raised when an unknown
+    string is passed as an optimizer.
+    """
+
+    X, y = make_classification()
+
+    optimizer = "sgf"  # sgf is not a loss
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        loss="binary_crossentropy",
+    )
+    with pytest.raises(
+        ValueError, match=r"Unknown optimizer:\s{0,1}" + f"{optimizer}"
+    ):
+        est.fit(X, y)
+
+
+def test_compiling_of_routed_parameters():
+    """Tests that routed parameters
+    can themselves be compiled.
+    """
+
+    X, y = make_classification()
+
+    class Foo:
+        got = dict()
+
+        def __init__(self, **kwargs):
```

Comment:
Suggested change:

``` diff
+ def __init__(self, **kwargs):
- def __init__(self, foo):
```

Only one keyword argument is passed, `foo`. Why not make it explicit?